proxy/src/schema/registry.rs

```diff
@@ -1,82 +1,149 @@
+        &Applied => match *self { TransactionState::Applied(ref a) => Some(a) },
+    }
+});
+
+/// Signals that [`Transaction`] has been successfully applied by a Node to a block. Carries the
+/// hash of the Block for further inspection and/or tracking of propagation.
+#[derive(GraphQLObject)]
+pub struct Applied {
+    /// Block hash the [`Transaction`] has been applied to.
+    pub block: juniper::ID,
+}
+
 /// Registry client wrapper.
 #[derive(Clone)]
-pub struct Registry<R>
-where
-    R: ClientT,
-{
+pub struct Registry {
     /// Registry client, whether an emulator or otherwise.
-    client: R,
+    client: Client,
 }
 
 /// Registry client wrapper methods
-impl<R> Registry<R>
-where
-    R: ClientT,
-{
-    // TODO(xla): Remvoe once integrated in the schema.
-    #[allow(dead_code)]
+impl Registry {
     /// Wrap a registry client.
-    pub fn new(client: R) -> Self {
+    pub const fn new(client: Client) -> Self {
         Self { client }
     }
 
-    // TODO(xla): Remvoe once integrated in the schema.
-    #[allow(dead_code)]
     /// Register a new project on the chain.
-    pub fn register_project(
+    pub async fn register_project(
         &self,
-        author: &Pair,
-        name: String,
+        author: &ed25519::Pair,
         domain: String,
-    ) -> Result<(), Error> {
+        name: String,
+    ) -> Result<H256, Error> {
+        // Verify that inputs are valid.
         let project_name = String32::from_string(name).map_err(ProjectValidation::NameTooLong)?;
         let project_domain =
             String32::from_string(domain).map_err(ProjectValidation::DomainTooLong)?;
-        let project_hash = H256::random();
+
+        // Prepare and submit checkpoint transaction.
+        let checkpoint_message = message::CreateCheckpoint {
+            project_hash: H256::random(),
+            previous_checkpoint_id: None,
+        };
+        let checkpoint_tx = registry::Transaction::new_signed(
+            author,
+            checkpoint_message,
+            TransactionExtra {
+                genesis_hash: self.client.genesis_hash(),
+                nonce: self.client.account_nonce(&author.public()).await?,
+            },
+        );
         let checkpoint_id = self
             .client
-            .sign_and_submit_call(
-                author,
-                CreateCheckpointParams {
-                    project_hash,
-                    previous_checkpoint_id: None,
-                },
-            )
-            // TODO(garbados): futurize
-            .wait()?
-            .wait()?
+            .submit_transaction(checkpoint_tx)
+            .await?
+            .await?
             .result?;
+
+        // Prepare and submit project registration transaction.
+        let register_message = message::RegisterProject {
+            id: (project_name, project_domain),
+            checkpoint_id,
+            metadata: registry::Bytes128::from_vec(vec![]).expect("unable construct metadata"),
```

Comment:
Why not use `?`?

Reply:
It's not meaningful to implement `From` for a `String` as it looses the context of the origin of the error. For that reason I created radicle-dev/radicle-registry#185.

Reply:
That makes sense.